Challenge1 - Setup iNope

ccjjmartin opened the pull request:
Prerequisites:
Aquifer is installed
Drupal VM is installed and configured for a site in your www directory with URL inope.drupalvm.dev

To setup up this site:

1) Change to your www directory:
cd ~/www

2) Clone this repository:
git clone git@github.com:ccjjmartin/aquifer-inope.git

3) Make a symlink for Drupal VM
ln -s aquifer-inope/build/ inope

4) Change into the new directory:
cd ~/www/aquifer-inope

5) Run
aquifer build

6) Your secret.settings.php file should be updated for your Drupal VM mysql setup something like:
$databases = array(
  'default' => array(
    'default' => array(
      'database' => 'inope',
      'username' => 'drupal',
      'password' => 'drupal',
      'host' => 'localhost',
      'port' => '',
      'driver' => 'mysql',
      'prefix' => '',
    ),
  ),
);

7) If using Drupal VM switch to your vm directory
cd ~/vm

8)  If VM is not running:
vagrant up

9) Connect to VM
vagrant ssh

10) Connect to mysql using your Drupal VM configuration setup something like:
mysql --host=localhost --user=root --password=root

11) In the mysql interface create a database named inope:
create database inope;

12) In the mysql interface grant all privileges on inope to the drupal user (or what you use in Drupal VM)
grant all on inope.\* to drupal@localhost;

13) Get your site alias something like @drupalvm.inope.drupalvm.dev:
drush sa

14) Install a new site:
drush @drupalvm.inope.drupalvm.dev site-install standard

15) Enable the master module
drush @drupalvm.inope.drupalvm.dev en master -y

16) Run the master execute command for the local scope
drush @drupalvm.inope.drupalvm.dev master-execute --scope=local

17) Log into your new site using:
drush @drupalvm.inope.drupalvm.dev uli 1

18) Create a new employee handbook at:
http://inope.drupalvm.dev/#overlay=node/add/inope-policy

Title: Employee Handbook
Body: Policies related to employees
Menu Link: Turn it on, it will default to the navigation menu

Save your new policy

18) Notice the review of employee handbook tab and click it

19) Try it out!

ccjjmartin committed:
Challenge 1 - iNope Initial Setup

Roles
anonymous role
authenticated role
staff role
administrator role

Content Type
Policy
--Tile
--Body
Editable by ONLY administrators
Viewable by ONLY staff and administrators

Custom module and features

inope
inope_policy_review
feature_inope_policy
feature_inope_site_configuration

Hook_menu for title
Title Callback

New Form
Editable at URL like node/%/review
--'I have reviewed this policy and approved it for use’ (default is checked)
--'I have reviewed this policy and would like to see changes'
--'Changes requested:' (#states (visible) if previous checkbox is true - required if above is true)
--Collect uid of user who submitted the form
--Which revision of the node they were viewing when feedback was submitted

hook_menu()
ONLY visible to staff and administrators
MENU_LOCAL_TASK

submission will only display a message with form values (sanitize it)

ccjjmartin committed:
Challenge 1 - iNope phpcs changes

ccjjmartin committed:
Challenge 1 - iNope Policy Review Module phpcs comments

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,121 @@
+ */
+function inope_policy_review_form($form, &$form_state) {
+
+  $form['inope_policy_review_approved_checkbox'] = array(
```
This is a bit weird. The states control over the checkbox is essentially making it behave like a single radio form control, unless there is something else you are doing here. Did you try a radio here instead and then make the other item conditional to the value selected?

thebruce replied:
Check this out:

http://www.nngroup.com/articles/checkboxes-vs-radio-buttons/

Idk about the quality of the article, honestly I think I learned form controls from visual basic when you were probably still in grade school, so I can't remember my trajectory to learn this stuff.  Anyway the important part is: 

```
    Radio buttons are used when there is a list of two or more options that are mutually exclusive and the user must select exactly one choice. In other words, clicking a non-selected radio button will deselect whatever other button was previously selected in the list.
    Checkboxes are used when there are lists of options and the user may select any number of choices, including zero, one, or several. In other words, each checkbox is independent of all other checkboxes in the list, so checking one box doesn't uncheck the others.
    A stand-alone checkbox is used for a single option that the user can turn on or off.
```

You will probably want to review HTML form controls -- you want to know when to use one vs the other, it does make a UX difference. If you need help finding some resources I can see what I can do -- but take a whack at finding some out there.

ccjjmartin replied:
I thought a radio control would be better.  Can’t explain why I stuck with checkboxes.  I will update it.

On Dec 16, 2015, at 4:59 PM, David Diers notifications@github.com wrote:

> In modules/custom/inope_policy_review/inope_policy_review.module:
> 
> > +/**
> > - \* The iNope Policy Review Page Title Callback.
> > - */
> >   +function inope_policy_review_tab_title_callback($node_nid) {
> > -  $node = node_load($node_nid);
> > -  $node_title = $node->title;
> > -  $title = t('Review of @node_title', array('@node_title' => $node_title));
> > -  return $title;
> >   +}
> >   +
> >   +/**
> > - \* The iNope Policy Review Form.
> > - */
> >   +function inope_policy_review_form($form, &$form_state) {
> >   +
> > -  $form['inope_policy_review_approved_checkbox'] = array(
> >   This is a bit weird. The states control over the checkbox is essentially making it behave like a single radio form control, unless there is something else you are doing here. Did you try a radio here instead and then make the other item conditional to the value selected?
> 
> —
> Reply to this email directly or view it on GitHub.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,121 @@
+/**
+ * The iNope Policy Review Page Title Callback.
+ */
+function inope_policy_review_tab_title_callback($node_nid) {
```
for nids in drupal you can probably assume that node isn't necessary.

If I had to do this i'd just go with $nid, its pretty common across core, so feel free to duplicate the pattern.  If you have some kind of weird arg -- you might want to be more specific. $uid is probably another one that you can just use.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,121 @@
+}
+
+/**
+ * The iNope Policy Review Page Title Callback.
```
Since this is not a hook and a custom function you probably need to utilize proper php doc style comments. That means you'll want to declare your params and returns in the comments.
https://www.drupal.org/node/1354#functions

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,121 @@
+function inope_policy_review_tab_title_callback($node_nid) {
+  $node = node_load($node_nid);
+  $node_title = $node->title;
+  $title = t('Review of @node_title', array('@node_title' => $node_title));
```
Good to see t('') here!

thebruce commented on drupal.make.yml:
```diff
@@ -25,12 +25,16 @@ projects:
     version: ~
   features_extra:
     version: ~
+  features_override:
```
Why this?  I can count the number of times i've ever needed to use features override and it worked on 1/2 finger.

thebruce replied:
What did you come up against that led you here?

ccjjmartin replied:
What is the correct module to use to determine the code differences in features?  I tried:

features_override
diff

thebruce replied:
diff - you mean from like an installed feature to overridden settings?
If so  - diff

thebruce commented on modules/custom/inope/inope.install:
```diff
@@ -0,0 +1,32 @@
+ * Enables the iNope Basic Configuration and Features.
+ */
+function inope_install() {
+  module_enable(array(
```
I need to check really quick - but I think you should be able to utilize the master module for these enables via aquifer. If so I'll report right back about that.

thebruce replied:
OK - i think so -- just make sure https://www.drupal.org/project/master is in your make file and then add the modules you'd like enabled to the settings.php file -- you should see the place, if not ping me.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,121 @@
+    'description' => 'Review of policy nodes.',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('inope_policy_review_form'),
+    'access arguments' => array('view published content'),
```
Not quite:
`Only staff and administrators should be able to see this form.`

Let's make this a specific permission, a custom permission to view this form. Let's grant that to staff and administrators and featurize that permission.  If this is new -- you'll need to look up how to create a custom permission on Drupal.org.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,121 @@
+ * The iNope Policy Review Form submit handler.
+ */
+function inope_policy_review_form_submit($form, &$form_state) {
+  $user = $GLOBALS['user'];
```
You actually want to be pretty careful with a variable called user.  I also think you don't want to get the user object in this way.

Look here: https://api.drupal.org/api/drupal/modules!user!user.module/function/user_access/7

See what they do, use this approach. Head to get user into $account as quickly as possible to protect yourself. $account is more abstract and a good way to avoid problems when dealing with all "user" objects.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,121 @@
+    'title arguments' => array(1),
+    'description' => 'Review of policy nodes.',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('inope_policy_review_form'),
```
OK - read this:
https://www.drupal.org/node/224170
This talks about wildcard loads.  See here that node/%/review is really a wildcard loader. With wildcard loaders you can have the value of the node passed in via page arguments, there is an example of how that's done.  So essentially you won't need to do a node load on your form. You could use this to set hidden values on your form or add to the form vars.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,121 @@
+ */
+function inope_policy_review_form_submit($form, &$form_state) {
+  $user = $GLOBALS['user'];
+  $node_id = array_slice(explode('/', check_plain($_GET['q'])), 1, 1);
```
again just go with $nid -- but really  might not need this if you pass it as a hidden. Again this is an internal facing form -- so any funny stuff -- is going to be sussed out and dealt with. Besides we need these values for processing later, so might as well have them here and use a validate function to make sure they are valid, right?

Also -- could you have used arg? https://api.drupal.org/api/drupal/includes!bootstrap.inc/function/arg/7

thebruce replied:
We also need the vid - so that is another thing we might pass along.

thebruce commented on modules/features/feature_inope_policy/feature_inope_policy.features.menu_links.inc:
```diff
@@ -0,0 +1,40 @@
+  $menu_links = array();
+
+  // Exported menu link: navigation_policy:node/add/inope-policy.
+  $menu_links['navigation_policy:node/add/inope-policy'] = array(
```
Woah --- why are we exporting node add menu items -- with the feature enabled aren't these there automatically?

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,121 @@
+      ),
+    ),
+  );
+
```
I would probably set the node nid and user id into the form here as hidden values

ccjjmartin committed:
Challenge 1 - Review Comment Change to Radio Form Input

ccjjmartin committed:
Challenge 1 - Add custom permission, update phpdoc block, update $node_nid

ccjjmartin committed:
Challenge 1 - Change $user to $account

ccjjmartin committed:
Challenge 1 - Nuked the add policy menu link

ccjjmartin committed:
Challenge 1 - Update form for nid, vid, uid and add validation

ccjjmartin committed:
Challenge 1 - Remove hook_install and replace with master file

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,166 @@
+    'title arguments' => array(1),
+    'description' => 'Review of policy nodes.',
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('inope_policy_review_form', 1),
```
I think to use the node load auto load you probably need a different page callback -- see that article i linked.  The idea is that you won't have to do a node load as a result it should already be there and loaded.

https://www.drupal.org/node/224170

Now take a further look there -- they actually give you some bad info, its not a trick, but you do have to read through -- we need a page callback so we can make use of the autoloader. We want the autoloader -- because in a way it sets you up for its own style of validation.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,166 @@
+      check_plain($form_state['values']['inope_policy_review_changes_requested_textarea'])) {
+    form_set_error('inope_policy_review_changes_requested_textarea', 'Please try again, only plain text in textarea');
+  }
+  if ($form_state['values']['inope_policy_review_hidden_uid'] != $GLOBALS['user']->uid) {
```
These are hidden values -- although you should perhaps validate them -- there is nothing to report to the user - besides maybe -- there has been an error.

What can a user really do with a hidden value?

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,166 @@
+ * The iNope Policy Review Form submit handler.
+ */
+function inope_policy_review_form_submit($form, &$form_state) {
+  $account = $GLOBALS['user'];
```
Still not there. 
Please look at the link i gave you:
Look here: https://api.drupal.org/api/drupal/modules!user!user.module/function/user_access/7
at the code specifically. 

This is best practice for creating a $account object from the user global.

thebruce replied:
And actually it might be best if this is coming from a user load from the uid from the form.  Of course you'll need to transfer the appropriate user account load there.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,166 @@
+
+  $form['inope_policy_review_hidden_uid'] = array(
+    '#type' => 'hidden',
+    '#value' => $GLOBALS['user']->uid,
```
This is still not the way to get the user object -- please see below.

thebruce commented on settings/settings.php:
```diff
@@ -25,19 +25,26 @@
     'comment',
+    'contextual',
+    'dashboard',
+    'dblog',
```
I don't see diff, that might be good.

thebruce commented on modules/features/feature_inope_policy/feature_inope_policy.features.user_permission.inc:
```diff
@@ -0,0 +1,59 @@
+function feature_inope_policy_user_default_permissions() {
+  $permissions = array();
+
+  // Exported permission: 'create inope_policy content'.
```
Did you export the permission for the inope form viewing anywhere?

thebruce commented:
A few more changes, get these in and I'll do a functional.

ccjjmartin committed:
Challenge 1 - Add autoloader for node in hook_menu().

ccjjmartin committed:
Challenge 1 - Change $GLOBALS['user'] to $user

ccjjmartin committed:
Challenge 1 - Define global $user

ccjjmartin committed:
Challenge 1 - Nuked validation on hidden inputs

ccjjmartin committed:
Challenge 1 - Switch the use of global $user to user_load().

ccjjmartin committed:
Challenge 1 - Add the diff module to the master list

ccjjmartin committed:
Challenge 1 - Export the view inope policy review permission

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,157 @@
+  
+  $form['inope_policy_review_hidden_nid'] = array(
+    '#type' => 'hidden',
+    '#value' => $form_state['build_info']['args'][0]->nid,
```
This is a bit opaque to the reader. I am sure you'd agree that $node->nid is a clearer statement of what we are setting the value to -- I think
$node = $form_state['build_info']['args'][0];
Would clear that up.

You can make it even more clear however by not using drupal get form as the callback and creating your own callback for the page -- which is what I was trying to get at. I think you thought I meant on the title callback, and its good you did what you did there -- to do that here you'd need a page callback where you pass the node object and then call drupal get form with an additional arg of the node.

This second half is like maybe not as important and I won't block functional on it -- but it's what I was expecting to see after my comments.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,157 @@
+    ),
+  );
+
+  $form['inope_policy_review_submit'] = array(
```
Just order-wise, and this might be my own preference -- I like submit at the bottom.

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,157 @@
+function inope_policy_review_form_validate($form, &$form_state) {
+  if ($form_state['values']['inope_policy_review_status_radio'] != 0 &&
+      $form_state['values']['inope_policy_review_status_radio'] != 1) {
+    form_set_error('inope_policy_review_status_radio', 'Please try again, status radio');
```
what would actually cause a radio button to not have either value if its a binary?  It's probably not human error - if i got this message and I had nothing to do with the issue - I would be hella frustrated.  "There has been a problem, please try again" - although less specific might be less frustrating. I don't know - put yourself in their position is "Please try again, status radio" - something that would make you feel good as a user?

thebruce commented on modules/custom/inope_policy_review/inope_policy_review.module:
```diff
@@ -0,0 +1,157 @@
+  }
+  if ($form_state['values']['inope_policy_review_changes_requested_textarea'] !=
+      check_plain($form_state['values']['inope_policy_review_changes_requested_textarea'])) {
+    form_set_error('inope_policy_review_changes_requested_textarea', 'Please try again, only plain text in textarea');
```
Again, -- what is the actual issue here -- what is it that isn't allowed -- what would cause the error to fire, give the user something that they can do something - with.  A user might come back with -- I copy and pasted from .... I don't understand what plain text is... etc.

thebruce commented:
One thing overall that I just noticed -- there are no inline comments in the code. It is good to document business logic in comments.  This example probably doesn't warrant the equivalent of moby dick in comments but there should be a few. The main idea is to make it simple for someone -- who might not know all the details of the feature to understand why you are doing what you are doing with the code.  Yes they could probably figure it out with enough time -- but who has that?  And if they do have that who wants to waste it trying to understand somebody's code from long ago.

thebruce commented:
You probably want to add aquifer coder https://github.com/aquifer/aquifer-coder to this project

thebruce commented:
Aquifer lint comments on this PR:

## FOUND 9 ERROR(S) AFFECTING 6 LINE(S)

  32 | ERROR | Last parameter comment requires a blank newline after it
  32 | ERROR | Doc comment for var $nid does not match actual variable name
     |       | $node at position 1
  32 | ERROR | Missing parameter type at position 1
  34 | ERROR | Data type of return value is missing
  34 | ERROR | Invalid @return data type, expected string but found String
  94 | ERROR | Whitespace found at end of line
  99 | ERROR | Whitespace found at end of line
 136 | ERROR | Inline comments must end in full-stops, exclamation marks, or
     |       | question marks
 150 | ERROR | Inline comments must end in full-stops, exclamation marks, or

##      |       | question marks

thebruce commented:
Did you mean to include conf settings for the master module in the settings file?  I don't currently see any.

Something like:
```
$conf['master_version'] = 2;
$conf['master_modules'] = array(
  'base' => array(

    // Core modules.
    'block',
    'file',
    'help',
    'list',
    'menu',
    'number',
    'options',
    'path',
    'url',
    // Contrib modules.
    'admin_menu',
    'addressfield',
    'features',
    'features_extra',
    'link',   
    'views',
    // Custom modules.
   
    // Feature modules.
  ),
  'local' => array(
    'dblog',
    'devel',
    'devel_generate',
    'diff',
    'feature_ws_developer',
    'field_ui',
    'module_filter',
    'views_ui',
  ),
);
```

ccjjmartin committed:
Challenge 1 - Change page callback in hook_menu()

Call back changed from drupal_get_form to inope_policy_review_page.
Also rearranged the order of the form items (submit button to bottom).

thebruce commented:
OK -- Functionally this checks out. So yay!

Let me know when you've had a chance to address this (hopefully) last round of feedback (minus the master settings stuff).

ccjjmartin committed:
Challenge 1 - Fix documentation for updated $node parameter

ccjjmartin committed:
Challenge 1 - Removed validation completely

ccjjmartin committed:
Challenge 1 - Add comments and update for phpcs review

ccjjmartin committed:
Challenge 1 - Remove unnecessary line of code

ccjjmartin committed:
Challenge 1 - Allow users to use include more than plain text in the textarea

ccjjmartin committed:
Challenge 1 - Show the user that the default value is approved
